### ISSUE TITLE
Can i directly read .rec file and use it on pytorch code without converting by numpy array?

### ISSUE THREAD

**Issue by Isaver23** [NVIDIA/DALI#796]
Can i directly read .rec file and use it on pytorch data format without converting by numpy array?

**Comment by Isaver23** [comment]
`python
tic = time.time()
for i, batch in enumerate(dali_train_iter):
    data, label = torch.from_numpy(batch[0].data[0].asnumpy()), torch.from_numpy(batch[0].label[0].asnumpy())
    if (i + 1) % args.disp_batches == 0:
        print('Batch [{}]\tSpeed: {:.2f} samples/sec'.format(i, args.disp_batches * args.batch_size / (time.time() - tic)))
        tic = time.time()
#     break
`

**Comment by JanuszL** [comment]
Hi,
Could you tell us more that is .rec file and what is inside it?

**Comment by Isaver23** [comment]
It's a classification dataset which is usually used by mxnet deep learning framework. but in fact I want to use the data inside the rec file to train a pytorch network. This is the code I used. 

```python

class HybridTrainPipe(Pipeline):
    def __init__(self, batch_size, num_threads, device_id, num_gpus):
        super(HybridTrainPipe, self).__init__(batch_size, num_threads, device_id, seed = 12 + device_id)
        self.input = ops.MXNetReader(path = [db_folder+"images.rec"], index_path=[db_folder+"images.idx"],
                                     random_shuffle = True, shard_id = device_id, num_shards = num_gpus)
        self.decode = ops.nvJPEGDecoderRandomCrop(device = "mixed",
                                                  output_type = types.RGB,
                                                  random_aspect_ratio = [0.8, 1.25],
                                                  random_area = [0.1, 1.0],
                                                  num_attempts = 100)
        self.resize = ops.Resize(device = "gpu", resize_x = 112, resize_y = 112)
        self.cmnp = ops.CropMirrorNormalize(device = "gpu",
                                            output_dtype = types.FLOAT,
                                            output_layout = types.NCHW,
                                            crop = (112, 112),
                                            image_type = types.RGB,
                                            mean = [0.485 * 255,0.456 * 255,0.406 * 255],
                                            std = [0.229 * 255,0.224 * 255,0.225 * 255])
        self.coin = ops.CoinFlip(probability = 0.5)

    def define_graph(self):
        rng = self.coin()
        self.jpegs, self.labels = self.input(name = "Reader")
        images = self.decode(self.jpegs)
        images = self.resize(images)
        output = self.cmnp(images, mirror = rng)
        return [output, self.labels]

trainpipes = [HybridTrainPipe(batch_size=batch_size, num_threads=T, device_id = i, num_gpus = N) for i in range(N)]
trainpipes[0].build()
dali_train_iter = DALIClassificationIterator(trainpipes, trainpipes[0].epoch_size("Reader"))

args = edict()
args.batch_size = batch_size
args.disp_batches = 500

import torch

tic = time.time()
for i, batch in enumerate(dali_train_iter):
    data, label = torch.from_numpy(batch[0].data[0].asnumpy()), torch.from_numpy(batch[0].label[0].asnumpy())
    if (i + 1) % args.disp_batches == 0:
        print('Batch [{}]\tSpeed: {:.2f} samples/sec'.format(i, args.disp_batches * args.batch_size / (time.time() - tic)))
        tic = time.time()
```
It works, but I think to thansfor from mxnet ndarray -> numpy array -> torch.tensor could make some unnecessary costs, so I'm thinking is there a better way to make it.

Thanks A LOT for your reply!

**Comment by JanuszL** [comment]
Hi,
You don't need to use numpy as DALi iterator in PyTorch returns torch tensors. The data source doesn't matter.
You can check https://docs.nvidia.com/deeplearning/sdk/dali-developer-guide/docs/examples/pytorch/pytorch-various-readers.html or even https://github.com/NVIDIA/DALI/blob/master/docs/examples/pytorch/resnet50/main.py.
That is true power of DALI, no matter what reader you are using data returned by the iterator still have the same format.

**Comment by Isaver23** [comment]
Your reply perfectly solved my problem! I'll check it soon and thank you SO MUCH again!

**Comment by JanuszL** [comment]
Hi,
If it doesn't work, please reopen. Closed now.
